### Protein language model rescue mutations highlight variant effects and structure in clinically relevant genes

This repo contains the scripts and metadata used in our work presented at NeurIPS 2022 Learning Meaningful Representation of Life (LMRL) workshop.

[Workshop website](https://www.lmrl.org/) | [Paper](link-to-arxiv-paper) | [Presentation](link-to-final-presentation)

### Pretrained models

| Model                                                      |    Number of layers | Number of parameters   |    Training dataset*    | Implementation in our work |
| ----------------------------------------------------------------------------------------------------- | :------------------: | :------------------: | :------------------: |:------------------: |
| [ESM-2](https://github.com/facebookresearch/esm#available-models)  | 33 | 650M | UR50/D | Single run |
| [ESM-1v](https://github.com/facebookresearch/esm#available-models) | 33 | 650M | UR90/S | Ensemble of 5 models|

* Briefly summarize the difference and its implication for our comparison.

### Data on genetic variation

| Description                                                      |    Data source | Notes   |  
| ----------------------------------------------------------------------------------------------------- | :------------------: | :------------------: |
| List of clinically actionable genes  | [ACMG v3.1](https://www.gimjournal.org/article/S1098-3600(21)05076-0/fulltext#secst0025) |  | 
| gnomad | xx | xx |
| clinvar | xx | xx |
| MSA | ucsc | xx |




## Citation

If you find this work useful, please cite it as follows:

```bibtex
@misc{,
  doi = {},
  url = {},
  author = {},
  keywords = {Machine Learning (cs.LG), Genomics (q-bio.GN), Applications (stat.AP), FOS: Computer and information sciences, FOS: Computer and information sciences, FOS: Biological sciences, FOS: Biological sciences},
  title = {},
  publisher = {arXiv},
  year = {2022},
  copyright = {Creative Commons Attribution 4.0 International}
}
```

## Feedback

If you have any questions or comments, or would like to collaborate, please feel free to reach out. 
